### PR TITLE
🐛 [RUMF-1369] Exclude error message from stacktrace parsing

### PR DESCRIPTION
**Pull request by liywjl**
## Motivation

We accidentally parse error messages that looked similar to stackTrace:

Example:
```
HttpError: 400 Bad Request, see context.apiResponse for more details
    at s (spa.min.js:2:2418592)
```
would output
```
    [{
      args: [],
      column: undefined,
      func: '?',
      line: undefined,
      url: 'HttpError: 400 Bad Request, see context.apiResponse for more details',
    }, {
      args: [],
      column: 5,
      func: 's',
      line: 3,
      url: 'spa.min.js',
    }]
```

`'HttpError: 400 Bad Request, see context.apiResponse for more details'` is not a valid URL

## Changes

Get the `toString` of the error (casting error as a `String`) and remove it from the beginning of the stack trace if present

## Testing

- [ ] Local
- [x] Staging
- [x] Unit
- [ ] End to end

---

I have gone over the [contributing](https://github.com/DataDog/browser-sdk/blob/main/CONTRIBUTING.md) documentation.

### CODE REVIEW

File: packages/core/src/domain/tracekit/computeStackTrace.ts [hunk 1]
```diff
@@ -104,9 +104,8 @@ const GECKO_EVAL_RE = /(\S+) line (\d+)(?: > eval line \d+)* > eval/i
 
 function parseGeckoLine(line: string): StackFrame | undefined {
   const parts = GECKO_LINE_RE.exec(line)
-  if (!parts) {
-    return
-  }
+  if (!parts) return
+  if (parts[3] === line && line !== '[native code]') return // overly greedy regex likely means this is a custom error message (exception native code)
```
Comment (BenoitZugmeyer) [inline packages/core/src/domain/tracekit/computeStackTrace.ts:108]: **💭 thought:** ‏ Ideally, the `parseGeckoLine` should not be used to parse React Native stacks :D anyway.


**💬 suggestion:** ‏Can I suggest an alternative fix? Let's be pragmatical here. The issue seems to be that:
* the error message is included in the `error.stack` property
* and the error message is considered as a stack "line" by our regexes.

Because the error message is often crafted by the application, this can happen unpredictibly. The message could literally be `    at namedFunc0 (http://example.com/js/script.js:10:2)` and we wouldn't be able to distinghuish it from a normal stack line. What we could do instead is, before trying to parse the `error.stack` value, test if it contains the error message and remove it from the stack if that's the case. Something like this:

```ts
let stackProperty = tryToGetString(ex, 'stack')
const exString = String(ex)
if (startsWith(stack, exString) {
  exProperty = stack.slice(exString.length)
}
stackProperty.split('\n').forEach(...)
```

WDYT?

Comment (liywjl) [reply packages/core/src/domain/tracekit/computeStackTrace.ts:108]: `parseGeckoLine` is currently catching a lot of cases unrelated to Gecko parsing errors (`'[native code]'` is just one of the examples). This is due to a greedy regex used (the main culprit is `.*?` just after `\[native)` in the `GECKO_LINE_RE`). 

Renaming or breaking down this part of the regex would be ideal to better understand what is getting matched.

Comment (liywjl) [reply packages/core/src/domain/tracekit/computeStackTrace.ts:108]: This is a great suggestion, I tried out something similar previously (but with different assumptions). 

The issue is that in some cases, the first line and the stack trace are the same, and this solution will inadvertently remove the string.

A lot of this argument is based on unit tests that failed. I suspect that despite it's flaws, `parseGeckoLine` does not regex match these edge cases.

Comment (BenoitZugmeyer) [reply packages/core/src/domain/tracekit/computeStackTrace.ts:108]: > The issue is that in some cases, the first line and the stack trace are the same, and this solution will inadvertently remove the string.

Which case? Do you have something in mind? I suspect that it would not happen very often, right? It would be really particular, hand crafted errors and stacks. And even if it happens, we would still send all the data (message + stack): even if the stack would miss its first line, it would be in the message anyway, so I think this is acceptable.

Comment (liywjl) [reply packages/core/src/domain/tracekit/computeStackTrace.ts:108]: 🤦‍♂️ `stack.slice` should have been `stackProperty.slice`. The solution you suggest actually solves this edge case without breaking anything else. Will update and add tests 

I still think we should maybe revisit `parseGeckoLine` at some point

File: packages/core/src/domain/tracekit/computeStackTrace.spec.ts [hunk 2]
```diff
@@ -499,6 +499,53 @@ Error: foo
...
+  it('should ignore custom error messages on chrome ', () => {
+    const stackFrames = computeStackTrace(new Error('bar@http://path/to/file.js:1:1'))
+
+    expect(stackFrames.stack[0]).not.toEqual({
+      args: [],
+      column: 1,
+      func: 'Error: bar',
+      line: 1,
+      url: 'http://path/to/file.js',
+    })
```
Comment (BenoitZugmeyer) [inline packages/core/src/domain/tracekit/computeStackTrace.spec.ts:511]: **💬 suggestion:** If anything change in the StackFrame format, this test will succeed without actually testing the expectation. Maybe only test for the URL?

```suggestion
    expect(stackFrames.stack[0].url).not.toBe('http://path/to/file.js')
```

File: packages/core/src/domain/tracekit/computeStackTrace.spec.ts [hunk 3]
```diff
@@ -499,6 +499,53 @@ Error: foo
...
+  })
+
+  it('should ignore messages that come from the `toString` of the error on chrome ', () => {
```
Comment (BenoitZugmeyer) [inline packages/core/src/domain/tracekit/computeStackTrace.spec.ts:514]: **💬 suggestion:** ‏this test case is about the same as the test case above. Maybe remove it?

File: packages/core/src/domain/tracekit/computeStackTrace.spec.ts [hunk 4]
```diff
@@ -499,6 +499,44 @@ Error: foo
...
+      line: 6,
+      url: 'spa.min.js',
+    })
+    expect(stackFrames.stack[2]).toEqual({
+      args: [],
+      column: 1,
+      func: '?',
+      line: 2,
+      url: 'spa.min.js',
+    })
```
Comment (liywjl) [inline packages/core/src/domain/tracekit/computeStackTrace.spec.ts:537]:
```suggestion
    const stackFrames = computeStackTrace(new Error('bar@http://path/to/file.js:1:1'))

    expect(stackFrames.stack[0].url).not.toBe('http://path/to/file.js')
```

**Review by BenoitZugmeyer — APPROVED** [review]

File: packages/core/src/domain/tracekit/computeStackTrace.spec.ts [hunk 5]
```diff
@@ -499,6 +499,12 @@ Error: foo
     })
   })
 
+  it('should ignore custom error messages ', () => {
```
Comment (bcaudan) [inline packages/core/src/domain/tracekit/computeStackTrace.spec.ts:502]: **💬 suggestion:** ‏Could you clarify test description?

**Review by bcaudan — APPROVED** [review]

File: packages/core/src/domain/tracekit/computeStackTrace.spec.ts (hunk 5 above)
Comment (liywjl) [inline packages/core/src/domain/tracekit/computeStackTrace.spec.ts:502]:
```suggestion
  it('should not include error message into stacktrace ', () => {
```
